alibi_detect/od/tests/test_knn.py

```diff
@@ -0,0 +1,184 @@
+import pytest
+import numpy as np
+import torch
+
+from alibi_detect.od.knn import KNN
+from alibi_detect.od.backend import AverageAggregatorTorch, TopKAggregatorTorch, MaxAggregatorTorch, \
+    MinAggregatorTorch, ShiftAndScaleNormalizerTorch, PValNormalizerTorch
+from sklearn.datasets import make_moons
+
+
+def make_knn_detector(k=5, aggregator=None, normalizer=None):
+    knn_detector = KNN(k=k, aggregator=aggregator, normalizer=normalizer)
+    x_ref = np.random.randn(100, 2)
+    knn_detector.fit(x_ref)
+    knn_detector.infer_threshold(x_ref, 0.1)
+    return knn_detector
+
+
+def test_unfitted_knn_single_score():
```

Comment:
The existing approach with drift detector tests is to test the backend detector on data, i.e. test [cd/pytorch/tests/test_mmd_pt.py](https://github.com/SeldonIO/alibi-detect/blob/master/alibi_detect/cd/pytorch/tests/test_mmd_pt.py) on real data in `cd/tests/test_mmd.py`, and then simply test that the backend-agnostic wrapper instantiates the correct backend (or raises an error) in [cd/tests/test_mmd.py](https://github.com/SeldonIO/alibi-detect/blob/master/alibi_detect/cd/tests/test_mmd.py).

The current implementation here changes this approach quite a bit by actually running tests on data in `test_knn.py`. Not saying this isn't a good idea, but would be good to consider what is the "ideal" approach since we are talking about rewriting tests soon.

Reply:
yeah, I've leant pretty heavily on integration style testing here, for instance in the `test_knn_ensemble_integration` and `test_knn_integration` tests. I feel like integration tests like this are quite important just to make sure all the pieces are in place. Appreciate this break from previous approaches. Perhaps I should reimplement the data-driven tests into the backend tests and refactor the integration tests into a separate test file as a best of both worlds?